File: spark-weather-rdd-exercise/src/main/scala/de/dimajix/training/spark/weather/Driver.scala
package de.dimajix.training.spark.weather

import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.slf4j.Logger
import org.slf4j.LoggerFactory

/**
  * Created by kaya on 03.12.15.
  */
object Driver {
  def main(args: Array[String]) : Unit = {
    // First create driver, so can already process arguments
    val options = new Options(args)
    val driver = new Driver(options)

    // Now create SparkContext (possibly flooding the console with logging information)
    val conf = new SparkConf()
      .setAppName("Spark Weather Analysis")
    val sc = new SparkContext(conf)

    // ... and run!
    driver.run(sc)
  }
}


class Driver(options:Options) {
  private val logger: Logger = LoggerFactory.getLogger(classOf[Driver])

  def run(sc: SparkContext) = {
    // 1. Load raw weather data from text file

    // 2. Transform raw data into meaningful RDD with WeatherData as object type

    // 3. Load raw station data from text file

    // 4. Transform raw data into meaningful RDD with StationData as object type

    // 5. Create new RDD from weather RDD with station code as key

    // 6. Create new RDD from station RDD with station code as key

    // 7. Join both RDDs

    // 8. Extract country, year as key and weather as value from joined RDD

    // 9. Aggregate results for getting min and max information. Have a look at aggregateByKey in PairRDD

    // 10. Save results as TextFile
  }
}

File: spark-python/weather-bcv-rdd/driver.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import optparse
import logging

from pyspark import SparkContext
from pyspark import SparkConf

from weather import StationData
from weather import WeatherData
from weather import WeatherMinMax


logger = logging.getLogger(__name__)



def create_context(appName):
    """
    Creates Spark HiveContext, with WebUI disabled and logging minimized
    """
    logger.info("Creating Spark context - may take some while")

    # Create SparkConf with UI disabled
    conf = SparkConf()
    conf.set("spark.hadoop.validateOutputSpecs", "false")

    sc = SparkContext(appName=appName, conf=conf)
    return sc


def parse_options():
    """
    Parses all command line options and returns an approprate Options object
    :return:
    """

    parser = optparse.OptionParser(description='PySpark Weather Analysis.')
    parser.add_option('-s', '--stations', action='store', nargs=1, help='Input file or directory containing station data')
    parser.add_option('-w', '--weather', action='store', nargs=1, help='Input file or directory containing weather data')
    parser.add_option('-o', '--output', action='store', nargs=1, help='Output file or directory')

    (opts, args) = parser.parse_args()

    return opts


def main():
    opts = parse_options()

    logger.info("Creating Spark Context")
    sc = create_context(appName="WordCount")

    logger.info("Starting processing")

    stations = sc.textFile(opts.stations) \
        .map(lambda line: StationData(line)) \
        .keyBy(lambda data: data.usaf + data.wban) \
        .collectAsMap()

    weather = sc.textFile(opts.weather) \
        .map(lambda line: WeatherData(line))

    weather_index = weather.keyBy(lambda data: data.usaf + data.wban)

    # joined_weather will contain tuples weather, station)
    #      [0] = weather
    #      [1] = station
    stations_bc = sc.broadcast(stations)
    joined_weather = weather_index.map(lambda (key,data): (data,stations_bc.value[key]))

    # Helper method for extracting (country, date) and weather
    def extract_country_year_weather(data):
        return ((data[1].country, data[0].date[0:4]), data[0])

    # Now extract country and year using the function above
    weather_per_country_and_year = \
        joined_weather.map(extract_country_year_weather)

    # Aggregate min/max information per year and country
    weather_minmax = weather_per_country_and_year \
        .aggregateByKey(WeatherMinMax(),lambda a,v:a.reduce(v), lambda l,r:l.combine(r))

    # Helper method for pretty printing
    def format_result(row):
        (k,v) = row
        country = k[0]
        year = k[1]
        minT = v.minTemperature or 0.0
        maxT = v.maxTemperature or 0.0
        minW = v.minWindSpeed or 0.0
        maxW = v.maxWindSpeed or 0.0
        line = "%s,%s,%f,%f,%f,%f" % (country, year, minT, maxT, minW, maxW)
        return line.encode('utf-8')

    # Store results
    weather_minmax \
        .map(format_result) \
        .saveAsTextFile(opts.output)

    logger.info("Successfully finished processing")


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    logging.getLogger('').setLevel(logging.INFO)

    logger.info("Starting main")
    main()
    logger.info("Successfully finished main")
